napari_plugin_devtools/validation.py

```diff
@@ -0,0 +1,126 @@
+import inspect
+import json
+import os
+from types import FunctionType
+from typing import Callable, Dict, List, Tuple, Union
+
+from napari.plugins import plugin_manager
+from napari_plugin_engine import HookImplementation
+from pkginfo import SDist, Wheel
+
+functions: Dict[Tuple[str, str], Callable] = dict()
+
+
+def validate_artifacts(folder):
+    """Validate the artifacts built for distribution.
+
+    Parameters
+    ----------
+    folder : str
+        artifacts folder after build with setuptools
+
+    Raises
+    ------
+    AssertionError
+        If the artifact is not build properly,
+    """
+    for pkgpath in os.listdir(folder):
+        pkgpath = os.path.join(folder, pkgpath)
+        if pkgpath.endswith('.tar.gz'):
+            dist = SDist(pkgpath)
+        elif pkgpath.endswith('.whl'):
+            dist = Wheel(pkgpath)
+        else:
+            print(f'Not a valid format {pkgpath}')
+            continue
+        assert 'Framework :: napari' in dist.classifiers
+        print(f'validated {pkgpath}')
+
+
+def validate_function(
+    args: Union[Callable, List[Callable]],
+    hookimpl: HookImplementation,
+):
+    """Validate the function is properly implemented in napari supported format
+
+    Parameters
+    ----------
+    args : Union[Callable, List[Callable]]
+        function to validate
+    hookimpl : HookImplementation
+        implementation of the hook
+
+    Raises
+    ------
+    AssertionError
+        When function is not in supported format
+    """
+    plugin_name = hookimpl.plugin_name
+    hook_name = '`napari_experimental_provide_function`'
+
+    assert plugin_name is not None, "No plugin name specified"
+    for func in args if isinstance(args, list) else [args]:
+        assert not isinstance(func, tuple), (
+            "To provide multiple function widgets "
+            "please use a LIST of callables"
+        )
+        assert isinstance(func, FunctionType), (
+            f'Plugin {plugin_name!r} provided a non-callable type to '
+            f'{hook_name}: {type(func)!r}. Function widget ignored.'
+        )
+
+        # Get function name
+        name = func.__name__.replace('_', ' ')
+
+        key = (plugin_name, name)
+        assert key not in functions, (
+            "Plugin '{}' has already registered a function widget '{}' "
+            "which has now been overwritten".format(*key)
+        )
+
+        functions[key] = func
+
+
+def list_function_implementations(plugin_name=None):
```

Comment:
We'll probably want a more general `find_hook_impls` at some point in time that will find and show you all the hook impls, and that could then be used by this more specific function oriented one